File: backend/app/schemas/instructor.py
from datetime import date, datetime
from pydantic import BaseModel

from app.models.course import CourseStatus


class InstructorCourseMetricsOut(BaseModel):
    course_id: str
    title: str
    status: CourseStatus
    enrolled_students: int
    completed_students: int
    quiz_average_score: float


class InstructorStudentRowOut(BaseModel):
    student_id: str
    student_name: str
    course_id: str
    course_title: str
    progress_pct: float
    last_access_at: datetime | None
    status: str


class StudentCourseProgressOut(BaseModel):
    course_id: str
    course_title: str
    progress_pct: float
    completed_at: datetime | None
    total_lessons: int
    completed_lessons: int


class InstructorStudentProgressDetailOut(BaseModel):
    student_id: str
    student_name: str
    courses: list[StudentCourseProgressOut]


class QuizScoreBinOut(BaseModel):
    label: str
    count: int


class InstructorQuizStatsOut(BaseModel):
    quiz_id: str
    quiz_title: str
    course_id: str
    course_title: str
    total_attempts: int
    pass_rate: float
    average_score: float
    score_distribution: list[QuizScoreBinOut]


class WeeklyEnrollmentPointOut(BaseModel):
    week_start: date
    enrollments: int


class MonthlyCompletionPointOut(BaseModel):
    month: str
    completions: int


class LessonWatchTimeLeaderOut(BaseModel):
    lesson_id: str
    lesson_title: str
    course_id: str
    course_title: str
    average_watch_time_sec: float


class InstructorAnalyticsOut(BaseModel):
    enrollments_by_week: list[WeeklyEnrollmentPointOut]
    completions_by_month: list[MonthlyCompletionPointOut]
    retention_rate: float
    top_watch_time_lesson: LessonWatchTimeLeaderOut | None

File: frontend/types/instructor.ts
import { CourseStatus } from "@/types/course";

export interface InstructorCourseMetrics {
  course_id: string;
  title: string;
  status: CourseStatus;
  enrolled_students: number;
  completed_students: number;
  quiz_average_score: number;
}

export interface InstructorStudentRow {
  student_id: string;
  student_name: string;
  course_id: string;
  course_title: string;
  progress_pct: number;
  last_access_at: string | null;
  status: "ACTIVE" | "INACTIVE" | "COMPLETED";
}

export interface StudentCourseProgress {
  course_id: string;
  course_title: string;
  progress_pct: number;
  completed_at: string | null;
  total_lessons: number;
  completed_lessons: number;
}

export interface InstructorStudentProgressDetail {
  student_id: string;
  student_name: string;
  courses: StudentCourseProgress[];
}

export interface QuizScoreBin {
  label: string;
  count: number;
}

export interface InstructorQuizStats {
  quiz_id: string;
  quiz_title: string;
  course_id: string;
  course_title: string;
  total_attempts: number;
  pass_rate: number;
  average_score: number;
  score_distribution: QuizScoreBin[];
}

export interface WeeklyEnrollmentPoint {
  week_start: string;
  enrollments: number;
}

export interface MonthlyCompletionPoint {
  month: string;
  completions: number;
}

export interface LessonWatchTimeLeader {
  lesson_id: string;
  lesson_title: string;
  course_id: string;
  course_title: string;
  average_watch_time_sec: number;
}

export interface InstructorAnalytics {
  enrollments_by_week: WeeklyEnrollmentPoint[];
  completions_by_month: MonthlyCompletionPoint[];
  retention_rate: number;
  top_watch_time_lesson: LessonWatchTimeLeader | null;
}
